pkg/lib/cockpit.js

```diff
@@ -1900,6 +1900,8 @@ function factory() {
             if (!options)
                 options = { };
             ensure_cache();
+            // this isn't a real class and only internal API; cockpit.d.ts annotates the return value DBusProxies
+            // @ts-expect-error TS7009: 'new' expression, whose target lacks a construct signature, implicitly has an 'any' type.
```

Comment:
So, I think the fundamental issue is that TypeScript doesn't recognize `function DBusProxies` as a constructor.  Things like
```
function Foo(x) {
  this.x = x;
}

const f = new Foo();
```
normally just work.

However, this produces the exact error TS7009:

```
function Bar(x) {
}

const b = new Bar();
```

The difference is that `Bar` does not assign to `this`, and neither does `DBusProxies`. It uses `Object.defineProperties`, and TS doesn't look inside of that apparently.

The TS7009 error goes away when adding a bogus `self.foo = 0` assignment inside `DBusProxies`. That is enough to make TS understand that it is a constructor.

I would prefer to do that. I think this will work:

```
        self.client = client;
        self.iface = iface;
        self.path_namespace = path_namespace;

        let waits;

        self.wait = function(func) {
            if (func)
                waits.always(func);
            return waits;
        }

        Object.defineProperties(self, {
            client: { enumerable: false, writable: false },
            iface: { enumerable: false, writable: false },
            path_namespace: { enumerable: false, writable: false },
            wait: { enumerable: false, writable: false }
        });
```

That is, we initialize the new object is the normal way, and only use Object.defineProperties to modify the "enumerable" and "writable" attributes of the properties.

Reply:
Niice, thanks for figuring this out! Reworked accordingly, let's confirm with the bo[ts]s!